lib/runtime/src/ops/topk.cc

```diff
@@ -42,396 +44,190 @@ using namespace FlexFlow::Kernels::TopK;
-  size_t batch_size = in1_domain.get_volume() / length;
-
-  forward_kernel_wrapper(
-      m, in_ptr, value_ptr, index_ptr, batch_size, length, k, m->sorted);
+static DeviceSpecific<TopKPerDeviceState>
+    init_task(Task const *task,
+              std::vector<PhysicalRegion> const &regions,
+              Context ctx,
+              Runtime *runtime) {
+  TaskArgumentAccessor acc(task, regions, ctx, runtime);
+  return init_task_impl(acc);
 }
 
-void TopK::backward(FFModel const &ff) {
-  ArgumentMap argmap;
-  Context ctx = ff.config.lg_ctx;
-  Runtime *runtime = ff.config.lg_hlr;
-  set_argumentmap_for_backward(ff, argmap);
-  IndexLauncher launcher(TOPK_BWD_TASK_ID,
-                         parallel_is,
-                         TaskArgument(NULL, 0),
-                         argmap,
-                         Predicate::TRUE_PRED,
-                         false /*must*/,
-                         0 /*mapper_id*/,
-                         outputs[0]->machine_view.hash());
-  // regions[0](I): value_grad
-  launcher.add_region_requirement(RegionRequirement(outputs[0]->part_grad,
-                                                    0 /*projection id*/,
-                                                    READ_ONLY,
-                                                    EXCLUSIVE,
-                                                    outputs[0]->region_grad));
-  launcher.add_field(0, FID_DATA);
-  // regions[1](I): indices
-  launcher.add_region_requirement(RegionRequirement(outputs[1]->part,
-                                                    0 /*projection id*/,
-                                                    READ_ONLY,
-                                                    EXCLUSIVE,
-                                                    outputs[1]->region));
-  launcher.add_field(1, FID_DATA);
-  // regions[2](I/O): input_grad
-  launcher.add_region_requirement(RegionRequirement(inputs[0]->part_grad,
-                                                    0 /*projection id*/,
-                                                    READ_WRITE,
-                                                    EXCLUSIVE,
-                                                    inputs[0]->region_grad));
-  launcher.add_field(2, FID_DATA);
-  runtime->execute_index_space(ctx, launcher);
+static optional<float> forward_task_impl(TaskArgumentAccessor const &acc) {
+  auto attrs = acc.get_argument<TopKAttrs>(ATTRS);
+  auto per_device_state =
+      acc.get_device_specific<TopKPerDeviceState>(PER_DEVICE_STATE);
+  auto profiling = acc.get_argument<ProfilingSettings>(PROFILING);
+
+  auto input = acc.get_tensor<Permissions::RO>(INPUT);
+  auto output = acc.get_tensor<Permissions::WO>(OUTPUT);
+
+  int length = input.shape.at(legion_dim_t(0)) + 1;
+  size_t batch_size = input.shape.get_volume() / length;
+  Allocator allocator = acc.get_allocator();
+  int *index_ptr = allocator.allocate(sizeof(int) * attrs.k);
```

Comment:
Since `INDICES` is an output of the operator, I think you should change this to `auto indices = acc.get_tensor<Permissions::WO>(INDICES);`

If you need to allocate anything, do that in the kernel then?